File: linuxdo-guest-browser/pycharm/src/test/java/studio/lexiao/linuxdo/ReaderHistoryTest.java
package studio.lexiao.linuxdo;

import java.util.ArrayList;
import java.util.List;

public final class ReaderHistoryTest {
    public static void main(String[] args) {
        roundTripAndDeduplicate();
        rejectPrivateAndDamagedEntries();
        capHistoryLength();
        updateLateTitleWithoutChangingVisitTime();
    }

    private static void roundTripAndDeduplicate() {
        ReaderHistory.Entry first = ReaderHistory.create("https://linux.do/t/example/123#post_2", "Example\nTopic", 1000);
        List<ReaderHistory.Entry> history = ReaderHistory.add(List.of(), first);
        history = ReaderHistory.add(history, ReaderHistory.create("https://linux.do/t/example/123", "Updated title", 2000));
        assert history.size() == 1;
        assert history.get(0).title().equals("Updated title");
        assert history.get(0).url().equals("https://linux.do/t/example/123");

        history = ReaderHistory.add(history, ReaderHistory.create("https://linux.do/t/example/123/10", "Floor ten", 3000));
        assert history.size() == 1;
        assert history.get(0).url().equals("https://linux.do/t/example/123");

        List<ReaderHistory.Entry> decoded = ReaderHistory.parse(ReaderHistory.serialize(history));
        assert decoded.equals(history);
    }

    private static void rejectPrivateAndDamagedEntries() {
        assert ReaderHistory.create("https://example.com/t/1", "off-site", 1000) == null;
        assert ReaderHistory.create("https://linux.do/login", "login", 1000) == null;
        assert ReaderHistory.create("https://user:pass@linux.do/latest", "credentials", 1000) == null;
        assert ReaderHistory.create("https://linux.do/latest?__cf_chl_tk=secret", "challenge", 1000) == null;
        assert ReaderHistory.create("https://linux.do/cdn-cgi/challenge-platform/test", "challenge", 1000) == null;
        assert ReaderHistory.parse("broken\tentry").isEmpty();
        assert ReaderHistory.parse("1000\t%%%\t%%%").isEmpty();
    }

    private static void capHistoryLength() {
        List<ReaderHistory.Entry> history = new ArrayList<>();
        for (int index = 0; index < 80; index += 1) {
            history = ReaderHistory.add(history, ReaderHistory.create(
                    "https://linux.do/t/topic/" + (index + 1),
                    "Topic " + index,
                    1000 + index
            ));
        }
        assert history.size() == ReaderHistory.MAX_ENTRIES;
        assert history.get(0).url().endsWith("/80");
    }

    private static void updateLateTitleWithoutChangingVisitTime() {
        ReaderHistory.Entry entry = ReaderHistory.create("https://linux.do/t/example/123", "公开页面", 4321);
        List<ReaderHistory.Entry> history = ReaderHistory.updateTitle(List.of(entry), "https://linux.do/t/example/123#post_2", "真实标题 - LINUX DO");
        assert history.size() == 1;
        assert history.get(0).title().equals("真实标题");
        assert history.get(0).visitedAt() == 4321;
    }
}

File: linuxdo-guest-browser/vscode/test/reader-history.test.js
'use strict';

const assert = require('node:assert/strict');
const test = require('node:test');
const {
  MAX_ENTRIES,
  addHistoryEntry,
  createHistoryEntry,
  normalizeStoredHistory
} = require('../src/reader-history');

test('history stores a public title, URL, visit time and reopen action', () => {
  const entry = createHistoryEntry(
    { type: 'topic', id: 123, slug: 'hello' },
    'Hello\nLINUX DO',
    'https://linux.do/t/hello/123#post_2',
    2000
  );
  assert.deepEqual(entry, {
    url: 'https://linux.do/t/hello/123',
    title: 'Hello LINUX DO',
    visitedAt: 2000,
    action: { type: 'topic', id: 123, slug: 'hello' }
  });
});

test('history rejects private/off-site entries and deduplicates by URL', () => {
  assert.equal(createHistoryEntry({ type: 'topic', id: 1, slug: 'one' }, 'One', 'https://example.com/t/one/1', 1), undefined);
  assert.equal(createHistoryEntry({ type: 'view', view: 'latest' }, 'Login', 'https://linux.do/login', 1), undefined);
  assert.equal(createHistoryEntry({ type: 'view', view: 'latest' }, 'Challenge', 'https://linux.do/latest?__cf_chl_tk=secret', 1), undefined);
  const first = createHistoryEntry({ type: 'topic', id: 1, slug: 'one' }, 'Old', undefined, 1);
  const latest = createHistoryEntry({ type: 'topic', id: 1, slug: 'one' }, 'New', 'https://linux.do/t/one/1/10', 2);
  assert.deepEqual(addHistoryEntry([first], latest), [latest]);
  assert.deepEqual(normalizeStoredHistory([{ broken: true }, latest]), [latest]);
});

test('history is capped at sixty entries', () => {
  let history = [];
  for (let id = 1; id <= 80; id += 1) {
    history = addHistoryEntry(history, createHistoryEntry({ type: 'topic', id, slug: `topic-${id}` }, `Topic ${id}`, undefined, id));
  }
  assert.equal(history.length, MAX_ENTRIES);
  assert.equal(history[0].action.id, 80);
});

test('history keeps a valid topic when its slug is missing or non-ASCII', () => {
  const missing = createHistoryEntry({ type: 'topic', id: 123 }, 'Missing slug', undefined, 1);
  const unicode = createHistoryEntry({ type: 'topic', id: 456, slug: '中文主题' }, 'Unicode slug', undefined, 2);
  assert.equal(missing.url, 'https://linux.do/t/topic/123');
  assert.deepEqual(missing.action, { type: 'topic', id: 123, slug: 'topic' });
  assert.equal(unicode.url, 'https://linux.do/t/topic/456');
});
